# Q1. Install the package

Once you installed the package, run the command mlflow --version and check the output.

What's the version that you have?

In [1]:
!mlflow --version

mlflow, version 2.3.2


Answer Q1: 2.3.2

# Q2. Download and preprocess the data

In [2]:
# download 2022 Green Taxi Trip Records (PARQUET)
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-01.parquet -P ./data
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-02.parquet -P ./data
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-03.parquet -P ./data

--2023-06-01 13:36:21--  https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-01.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 13.33.244.164, 13.33.244.13, 13.33.244.127, ...
Connecting to d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)|13.33.244.164|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1254291 (1.2M) [binary/octet-stream]
Saving to: ‘./data/green_tripdata_2022-01.parquet’

green_tripdata_2022 100%[===================>]   1.20M  1.02MB/s    in 1.2s    

2023-06-01 13:36:23 (1.02 MB/s) - ‘./data/green_tripdata_2022-01.parquet’ saved [1254291/1254291]

--2023-06-01 13:36:23--  https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-02.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 13.33.244.91, 13.33.244.127, 13.33.244.13, ...
Connecting to d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)|13.33.244.91|:443... connected.
HTTP r

In [3]:
!python preprocess_data.py --raw_data_path data --dest_path ./output

In [8]:
! du -h output/*

152K	output/dv.pkl
2.6M	output/test.pkl
2.1M	output/train.pkl
2.3M	output/val.pkl


Answer Q2:
```
152K	output/dv.pkl
```

# Q3. Train a model with autolog

In [9]:
!python train.py

2023/06/01 13:46:59 INFO mlflow.tracking.fluent: Experiment with name 'nyc-taxi-homework' does not exist. Creating a new experiment.
2023/06/01 13:47:05 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/home/admin-gpu/miniconda3/envs/mlops-zoom/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils."


MlFlows logging models parameters
```
max_depth	10
```

Answer Q3: 10

# Q4. Tune model hyperparameters

In [11]:
!python hpo.py

2023/06/01 14:08:43 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-hyperopt' does not exist. Creating a new experiment.
[I 2023-06-01 14:08:43,987] A new study created in memory with name: no-name-9766fbc8-deae-44ab-8170-ee6e77e7c336
[I 2023-06-01 14:08:44,576] Trial 0 finished with value: 2.451379690825458 and parameters: {'n_estimators': 25, 'max_depth': 20, 'min_samples_split': 8, 'min_samples_leaf': 3}. Best is trial 0 with value: 2.451379690825458.
[I 2023-06-01 14:08:44,730] Trial 1 finished with value: 2.4667366020368333 and parameters: {'n_estimators': 16, 'max_depth': 4, 'min_samples_split': 2, 'min_samples_leaf': 4}. Best is trial 0 with value: 2.451379690825458.
[I 2023-06-01 14:08:45,211] Trial 2 finished with value: 2.449827329704216 and parameters: {'n_estimators': 34, 'max_depth': 15, 'min_samples_split': 2, 'min_samples_leaf': 4}. Best is trial 2 with value: 2.449827329704216.
[I 2023-06-01 14:08:45,522] Trial 3 finished with value: 2.460983516558473 a

![random-forest-hyperopt](repo_pics/random-forest-hyperopt.png)

Answer Q4: 2.45

# Q5. Promote the best model to the model registry

In [16]:
!python register_model.py

2023/06/01 14:40:49 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-best-models' does not exist. Creating a new experiment.
2023/06/01 14:40:51 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/home/admin-gpu/miniconda3/envs/mlops-zoom/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils."
run id: 49831e58292c49ca986880d900eae85b, rmse: 2.2855
Successfully registered model 'random-forest_best_model'.
2023/06/01 14:40:58 INFO mlflow.tracking._model_registry.client: Waiting up to 300 seconds for model version to finish creation. Model name: random-forest_best_model, version 1
Created version '1' of model 'random-forest_best_model'.


Answer Q5: 2.2855

# Q6. Model metadata

Answer Q5: Version number